x-pack/plugins/security_solution/server/lib/detection_engine/prebuilt_rules/logic/diff/calculation/algorithms/scalar_array_diff_algorithm.ts

```diff
@@ -36,75 +37,81 @@ export const scalarArrayDiffAlgorithm = <TValue>(
-    baseVersion,
+  const hasBaseVersion = baseVersion !== MissingVersion;
+
+  const { mergeOutcome, conflict, mergedVersion } = mergeVersions({
+    hasBaseVersion,
+    baseVersion: hasBaseVersion ? baseVersion : undefined,
     currentVersion,
     targetVersion,
     diffOutcome,
   });
 
   return {
-    base_version: baseVersion,
+    has_base_version: hasBaseVersion,
+    base_version: hasBaseVersion ? baseVersion : undefined,
     current_version: currentVersion,
     target_version: targetVersion,
     merged_version: mergedVersion,
+    merge_outcome: mergeOutcome,
 
     diff_outcome: diffOutcome,
-    merge_outcome: mergeOutcome,
+    conflict,
     has_update: valueCanUpdate,
-    has_conflict: mergeOutcome === ThreeWayMergeOutcome.Conflict,
   };
 };
 
 interface MergeResult<TValue> {
   mergeOutcome: ThreeWayMergeOutcome;
   mergedVersion: TValue[];
+  conflict: ThreeWayDiffConflict;
 }
 
 interface MergeArgs<TValue> {
-  baseVersion: TValue[] | MissingVersion;
+  hasBaseVersion: boolean;
+  baseVersion: TValue[] | undefined;
   currentVersion: TValue[];
   targetVersion: TValue[];
   diffOutcome: ThreeWayDiffOutcome;
 }
 
 const mergeVersions = <TValue>({
+  hasBaseVersion,
   baseVersion,
   currentVersion,
   targetVersion,
   diffOutcome,
 }: MergeArgs<TValue>): MergeResult<TValue> => {
-  const dedupedBaseVersion = baseVersion !== MissingVersion ? uniq(baseVersion) : MissingVersion;
+  const dedupedBaseVersion = uniq(baseVersion);
   const dedupedCurrentVersion = uniq(currentVersion);
   const dedupedTargetVersion = uniq(targetVersion);
 
   switch (diffOutcome) {
+    // Scenario -AA is treated as scenario AAA:
+    // https://github.com/elastic/kibana/pull/184889#discussion_r1636421293
+    case ThreeWayDiffOutcome.MissingBaseNoUpdate:
     case ThreeWayDiffOutcome.StockValueNoUpdate:
     case ThreeWayDiffOutcome.CustomizedValueNoUpdate:
-    case ThreeWayDiffOutcome.CustomizedValueSameUpdate: {
+    case ThreeWayDiffOutcome.CustomizedValueSameUpdate:
       return {
-        mergeOutcome: ThreeWayMergeOutcome.Current,
+        conflict: ThreeWayDiffConflict.NONE,
         mergedVersion: dedupedCurrentVersion,
+        mergeOutcome: ThreeWayMergeOutcome.Current,
       };
-    }
+
     case ThreeWayDiffOutcome.StockValueCanUpdate: {
       return {
-        mergeOutcome: ThreeWayMergeOutcome.Target,
+        conflict: ThreeWayDiffConflict.NONE,
         mergedVersion: dedupedTargetVersion,
+        mergeOutcome: ThreeWayMergeOutcome.Target,
       };
     }
+
     case ThreeWayDiffOutcome.CustomizedValueCanUpdate: {
-      if (dedupedBaseVersion === MissingVersion) {
-        return {
-          mergeOutcome: ThreeWayMergeOutcome.Merged,
-          mergedVersion: union(currentVersion, targetVersion),
-        };
-      }
-
-      const addedCurrent = difference(dedupedCurrentVersion, dedupedBaseVersion);
+      const addedCurrent = difference(dedupedCurrentVersion, dedupedBaseVersion as TValue[]);
```

Comment:
We always try to avoid `as` casting in production code, can we try to rewrite this algorithm in a type-safe way?

Reply:
Fixed